Automattic/simplenote-ios
# Snackbar: First steps at potential implementation

### Details
Part of #938, **In-App Notifications: Snackbar**. This is the initial plumbing demonstrating one possible way to present Snackbars. Covers the view lifecycle, display, and clean up. Basic support for dynamic text.

Still needs:

- [x] 1. Proper theming for light & dark mode.
- [x] 2. Re-sizing the view based on message length. The original Issue shows the containing view only taking up the width necessary to show the message + the action button's title. Need some clarity on clipping vs wrapping longer messages.
- [x] 3. Additional initializer to wire up the action button. Currently the button just logs when clicked. Additional initializer would include a closure and execute that when clicked.
- [x] 4. Prevents additional notices if one is currently displayed.
- [x] 5. Needs to set duration based on presence of action button. Currently just puts up the notice for 2 seconds regardless of type.
- [ ] 6. Final colours for Dark & Light mode.
- [ ] 7. Haptic feedback.
- [ ] 8. Rotation support (appears to work but not throughly tested yet).
- [ ] 9. Remove debug logs showing presentation flow.

Not implemented but might be nice to have:

- [ ] Queuing mechanism if we wanted to chain notices.

### Test
1. Build and run.
2. With a note visible, click the Note Info button to see a simple message. Click the Note Options button to see a message with an action button.
3. See the Snackbar animate in then fade out.

### Release
**Do not merge!** This is a draft of work in progress.

### Comments

The basic architecture is built around the `Snackbar` and `SnackbarPresenter` classes with `SnackbarViewController` for XIB-based loading of the view. Only the `Snackbar` class is needed to present a notification.

```
let snackbar = Snackbar(message: "Some message")
snackbar.show()
```

For a Snackbar with action button:
```
let snackbar = Snackbar(message: "Snackbar with action!", actionTitle: "Boom!") {
    print("Callback code executed!")
}

```

**Light Mode example:**

<img src="https://user-images.githubusercontent.com/40267301/99146734-b7d41e00-26b5-11eb-96d8-41dbfa9c4027.gif" width="300" />

**Dark Mode with larger text:**

<img src="https://user-images.githubusercontent.com/40267301/99146778-fcf85000-26b5-11eb-9f07-8ac80d6ce91e.gif" width="300" />
